source/backends/ubuntu/ubupkg.d

```diff
@@ -0,0 +1,167 @@
+ * Author: Iain Lane <iain.lane@canonical.com>
+ *
+ * Licensed under the GNU Lesser General Public License Version 3
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the license, or
+ * (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this software.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+module backends.ubuntu.ubupkg;
+
+import appstream.Component;
+
+import backends.debian.debpkg;
+import bindings.gettext;
+import backends.interfaces;
+
+import glib.Internationalization;
+import glib.KeyFile;
+
+import std.container : Array;
+
+import logging;
+
+immutable DESKTOP_GROUP = "Desktop Entry";
+
+class UbuntuPackage : DebPackage
+{
+    this (string pname, string pver, string parch, string globalTmpDir, ref Array!Package allPackages)
+    {
+        this.globalTmpDir = globalTmpDir;
+        this.langpackDir = buildPath (globalTmpDir, "langpacks");
+        this.localeDir = buildPath (langpackDir, "locales");
+        this.allPackages = allPackages;
+        super (pname, pver, parch);
+    }
+
+    override string[string] processDesktopFile (KeyFile desktopFile, const string text)
+    {
+        string langpackdomain;
+
+        try {
+            langpackdomain = desktopFile.getString (DESKTOP_GROUP,
+                                                    "X-Ubuntu-Gettext-Domain");
+            logDebug ("%s has langpack domain %s", name, langpackdomain);
+
+            synchronized {
+                extractLangpacks ();
+                return getTranslations (langpackdomain, text);
+            }
+        } catch {
+            return null;
+        }
+    }
+
+private:
+    string globalTmpDir;
+    string langpackDir;
+    string localeDir;
+    string[] langpackLocales;
+    Array!Package allPackages;
+
+    void extractLangpacks ()
+    {
+        import std.algorithm : filter, map;
+        import std.array : appender, array, split;
+        import std.file : dirEntries, exists, SpanMode, readText;
+        import std.parallelism : parallel;
+        import std.path : baseName;
+        import std.process : Pid, spawnProcess, wait;
+        import std.string : splitLines, startsWith;
+
+        auto path = buildPath (langpackDir, "usr", "share", "locale-langpack");
+
+        if (!langpackDir.exists) {
+            foreach (pkg; allPackages) {
+                if (!pkg.name.startsWith ("language-pack"))
+                    continue;
+
+                /* :( */
+                UbuntuPackage upkg = cast (UbuntuPackage) pkg;
```

Comment:
Jap - casting is okay-ish, but ideally use `to!UbuntuPackage (pkg)` instead of a direct cast. (`to` does a bit of compile-time error checking).
https://dlang.org/phobos/std_conv.html#.to